security/pkg/stsservice/server/server.go

```diff
@@ -109,6 +109,21 @@ func (s *Server) ServeStsRequests(w http.ResponseWriter, req *http.Request) {
 		s.sendErrorResponse(w, invalidRequest, validationError)
 		return
 	}
+	// Authentication using the original JWT - Google requests use a Scope.
+	// This is used for authenticating directly with the token, for XDS/Citadel
+	if reqParam.Scope == "" {
```

Comment:
Checking "scope" is kind of hacking. The right way is to specify token manager plugin type (instead of "GoogleTokenExchange", use "NoExchange").

Reply:
That's not possible - we need both GoogleTokenExchange and NoExchange at the same time. 

Stackdriver for example will pass a scope - and go the google exchange. 
We may have other STS exchanges as well in the future, for example to generate K8S tokens with 
a different audience. 

One option is to make the decision based on audience - but it's still tricky. 

Since this is temporary - the XDS proxy in agent doesn't need to use STS - I think checking for scope is good enough for now, but if you have another recommendation that allows both google and no exchange I can change.

Reply:
so what is the long term solution if this is temporary?

Reply:
Istio-agent will handle the connection to Istiod - and envoy connects to a UDS socket to istio-agent.

In istio-agent we can have explicit code to get the token, without going trough hoops.

Reply:
BTW - we may still want STS to be able to support returning K8S JWTs with audience for other purposes than connecting to the XDS server. So having an STS server that can do more than just google token is great - but not the purpose of this PR.

Reply:
@costinm Is this feature going to let Envoy send STS request to the STS agent for JWT authentication only, and then use the JWT returned from STS agent to set up XDS connection with Istiod?

Can you also add a unit test for this change in server_test.go?

Reply:
@liminw for VM case, envoy will send VM credential to STS agent for token exchange, it is not clear whether VM credential from different cloud providers always set scope. Can we let Istiod do the JWT validation instead of STS agent?

Reply:
Yes, this lets envoy send STS requests to the agent, and then connect to Istiod. Right now it's just returning the current token. 

Testing is currently happening in the cloudrun ( since I'm using this for getting the CR variant of Istio install), will
be extensively covered since that's the only auth method supported.  I can also add a mock test, but unlikely to help
much, most of the issues so far have been around integration.

Istiod is now doing the JWT validation. For VMs - it'll be a bit more complicated, this PR is not covering that yet.

Reply:
@costinm thanks for the information! Are you using this bootstrap config [template](https://github.com/istio/istio/blob/master/tools/packaging/common/gcp_envoy_bootstrap.json#L165) ? If so, then we can set a new path [here](https://github.com/istio/istio/blob/0e8aebca29dc1dd0df1af01e5701176f39fb7d43/tools/packaging/common/gcp_envoy_bootstrap.json#L165)  e.g. `http://localhost:{{ .sts_port }}/echo`, we can register a new function for `/echo` [here](https://github.com/istio/istio/blob/0e8aebca29dc1dd0df1af01e5701176f39fb7d43/security/pkg/stsservice/server/server.go#L80). This change is also small and easier to maintain.  @liminw

Reply:
@JimmyCYJ VM is using the same setting (scope, etc) when talking to Google services. 

Jimmy's suggestion sounds like a good idea. Using a different ""token_exchange_service_uri" make it more explicit that it is for a different purpose.